examples/mnist/Main.hs

```diff
@@ -0,0 +1,159 @@
+
+--------------------------------------------------------------------------------
+-- Multi-Layer Perceptron (MLP)
+--------------------------------------------------------------------------------
+
+
+newtype Parameter dtype shape = Parameter A.IndependentTensor deriving (Show)
+
+toDependent :: Parameter dtype shape -> Tensor dtype shape
+toDependent (Parameter t) = UnsafeMkTensor $ A.toDependent t
+
+instance A.Parameterized (Parameter dtype shape) where
+  flattenParameters (Parameter x) = [x]
+  replaceOwnParameters _ = Parameter <$> A.nextParameter
+
+data LinearSpec (dtype :: D.DType) (inputFeatures :: Nat) (outputFeatures :: Nat) = LinearSpec
+  deriving (Show, Eq)
+
+data Linear (dtype :: D.DType) (inputFeatures :: Nat) (outputFeatures :: Nat) =
+  Linear { weight :: Parameter dtype '[inputFeatures, outputFeatures]
+         , bias :: Parameter dtype '[outputFeatures]
+         } deriving (Show, Generic)
+
+linear
+  :: Linear dtype inputFeatures outputFeatures
+  -> Tensor dtype '[batchSize, inputFeatures]
+  -> Tensor dtype '[batchSize, outputFeatures]
+linear Linear {..} input =
+  add (mm input (toDependent weight)) (toDependent bias)
+
+makeIndependent :: Tensor dtype shape -> IO (Parameter dtype shape)
+makeIndependent t = Parameter <$> A.makeIndependent (toDynamic t)
+
+instance A.Parameterized (Linear dtype inputFeatures outputFeatures)
+
+instance (KnownDType dtype, KnownNat inputFeatures, KnownNat outputFeatures) => A.Randomizable (LinearSpec dtype inputFeatures outputFeatures) (Linear dtype inputFeatures outputFeatures) where
+  sample LinearSpec =
+    Linear <$> (makeIndependent =<< randn) <*> (makeIndependent =<< randn)
+
+data MLPSpec (dtype :: D.DType) (inputFeatures :: Nat) (outputFeatures :: Nat) (hiddenFeatures :: Nat) = MLPSpec
+
+data MLP (dtype :: D.DType) (inputFeatures :: Nat) (outputFeatures :: Nat) (hiddenFeatures :: Nat) =
+  MLP { layer0 :: Linear dtype inputFeatures hiddenFeatures
+      , layer1 :: Linear dtype hiddenFeatures hiddenFeatures
+      , layer2 :: Linear dtype hiddenFeatures outputFeatures
+      } deriving (Show, Generic)
+
+instance A.Parameterized (MLP dtype inputFeatures outputFeatures hiddenFeatures)
+
+instance (KnownDType dtype, KnownNat inputFeatures, KnownNat outputFeatures, KnownNat hiddenFeatures) => A.Randomizable (MLPSpec dtype inputFeatures outputFeatures hiddenFeatures) (MLP dtype inputFeatures outputFeatures hiddenFeatures) where
+  sample MLPSpec =
+    MLP <$> A.sample LinearSpec <*> A.sample LinearSpec <*> A.sample LinearSpec
+
+mlp
+  :: MLP dtype inputFeatures outputFeatures hiddenFeatures
+  -> Tensor dtype '[batchSize, inputFeatures]
+  -> Tensor dtype '[batchSize, outputFeatures]
+mlp MLP {..} = linear layer2 . relu . linear layer1 . relu . linear layer0
+
+model
+  :: MLP dtype inputFeatures outputFeatures hiddenFeatures
+  -> Tensor dtype '[batchSize, inputFeatures]
+  -> Tensor dtype '[batchSize, outputFeatures]
+model = (sigmoid .) . mlp
```

Comment:
Given how trivial this is, I'd be in favor of merging mlp into model.

Reply:
standard thing to do is to use log softmax rather than sigmoid for multiclass classification, see the pytorch reference example https://github.com/pytorch/examples/blob/master/mnist/main.py

Reply:
@austinvhuang @junjihashimoto you are completely right, I’m not sure why I didn’t see this issue when I looked at the code. Sigmoid won’t work here, it has to be a softmax.